Dataview set/get negative index throw RangeError Fixes: #4978

rhuanjl opened the pull request:
Simple fix for #4978 

Note the ToIndex method I used was based on the implementation in ArrayBuffer.cpp but slightly simpler as DataView.cpp has a check for lengths that are too big later on.

Could clean up the code slightly by using a consistent version of ToIndex in both places though that would be a more invasive change than this.

jackhorton commented on lib/Runtime/Library/DataView.h:
```diff
@@ -55,6 +55,34 @@ namespace Js
             return static_cast<DataView*>(aValue);
         }
 
+        static uint32 ToIndex (Var value, int32 errorCode, ScriptContext *scriptContext)
```
Especially since this is currently specific to DataView, shouldn't this method not take an error code and always throw InvalidOffset?

akroshg commented on lib/Runtime/Library/DataView.h:
```diff
@@ -55,6 +55,34 @@ namespace Js
             return static_cast<DataView*>(aValue);
         }
 
+        static uint32 ToIndex (Var value, ScriptContext *scriptContext)
```
How is it different than ArrayBuffer::ToIndex ?

akroshg replied:
Oh you mentioned that in the description that they are almost similar. Good to have one of them.

---
In reply to: [181444808](https://github.com/Microsoft/ChakraCore/pull/4981#discussion_r181444808) [](ancestors = 181444808)

rhuanjl replied:
Do you want me to refactor this to share one method or leave as is?

dilijev replied:
If it is straightforward to capture the difference, better to refactor. To avoid compiling in unnecessary branches (since the actual method needed in each case is known to us at compile-time) you can use a `template <bool SomeConditional>` strategy.

akroshg commented:
Thanks for fixing this - looks good to me.

rhuanjl commented:
I've added a separate commit to refactor this to use ArrayBuffer::ToIndex and lose the new method.

As that method include a length check I've removed the length checks from the GetValue and SetValue methods - this seemed the neatest way to implement this noting that the GetValue and SetValue methods are only used within DataView.cpp and are always now preceded immediately by a call to ToIndex.

sethbrenith commented on lib/Runtime/Library/DataView.h:
```diff
@@ -104,106 +104,91 @@ namespace Js
                 JavascriptError::ThrowTypeError(scriptContext, JSERR_DetachedTypedArray, funcName);
             }
-            if ((byteOffset + sizeof(TypeName) <= GetLength()) && (byteOffset <= GetLength()))
+            uint32 byteOffset = ArrayBuffer::ToIndex(offset, JSERR_DataView_InvalidOffset, scriptContext, GetLength() - sizeof(TypeName), false);
```
>GetLength() - sizeof(TypeName) [](start = 106, length = 30)

Seems like we risk an underflow here if length is very small. Let's add a test case that makes sure we throw if the length is zero and we request element zero.

rhuanjl replied:
May need to specifically check for that actually good shout.

jackhorton commented:
@dotnet-bot test OSX static_osx_osx_debug please
test OSX static_osx_osx_release please
test Ubuntu static_ubuntu_linux_debug please

rhuanjl commented on lib/Runtime/Library/DataView.cpp:
```diff
@@ -159,8 +159,7 @@ namespace Js
         DataView* dataView = DataView::FromVar(args[0]);
-        uint32 offset = JavascriptConversion::ToUInt32(args[1], scriptContext);
-        return dataView->template GetValue<int8>(offset, _u("DataView.prototype.GetInt8"), FALSE);
+        return dataView->template GetValue<int8>(args[1], _u("DataView.prototype.GetInt8"), FALSE);
```
Assuming you mean the one here and the one for GetValue uint16 I couldn't think of any purpose for them but they seem to have been added intentionally in this commit: 13f6e3b54a057ed8954fe55a2decb27eeae0fe3e

So I was hesitant to remove them.

jackhorton replied:
Interesting... I have no idea why those would be necessary, but alright. Theyre not hurting anyone, I suppose.

jackhorton commented:
Assuming no one else has comments here, I can merge this today.

rhuanjl commented:
@jackhorton thanks, let me know if you want me to update and/or squash the two commits.

(I was keeping them separate in case anyone didn't like the way I'd refactored this)

jackhorton commented:
Since we always do no-ff merges and walk first-parent history in all of our internal tooling, ive never had strong feelings on the rule about squashing. I think 2 commits is fine.

jackhorton commented:
Update on this, we are having internal infra issues. Haven't forgotten about it!
